### PR TITLE
docs(recipes): Snapshots

### PR DESCRIPTION
**Pull request by igorlukanin**
Please review the new recipe for the data snapshots (SCD type 2).

### CODE REVIEW

**Comment by igorlukanin** [comment]
@hassankhan I wonder why Netlify wasn't fired for the docs changes in this PR...

File: docs/content/Examples-Tutorials-Recipes/Snapshots.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+<!-- prettier-ignore-start -->
+[[info | ]]
+| To generate a range of dates, here we use the
+| [`GENERATE_SERIES` function](https://www.postgresql.org/docs/9.1/functions-srf.html)
+| which is Postgres-specific. Other databases have similar functions, e.g.,
+| [`GENERATE_DATE_ARRAY`](https://cloud.google.com/bigquery/docs/reference/standard-sql/array_functions#generate_date_array)
+| in BigQuery.
+<!-- prettier-ignore-end -->
+
+Please note that it makes sense to make the `StatusSnapshots` cube to
```
Comment (hassankhan) [inline docs/content/Examples-Tutorials-Recipes/Snapshots.md:115]:
```suggestion
Please note that it makes sense to make the `StatusSnapshots` cube
```

File: docs/content/Examples-Tutorials-Recipes/Snapshots.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+[extend](https://cube.dev/docs/schema/reference/cube#parameters-extends)
```
Comment (hassankhan) [inline docs/content/Examples-Tutorials-Recipes/Snapshots.md:116]:
```suggestion
[extend](https://cube.dev/docs/schema/reference/cube#parameters-extends) from
```

File: docs/content/Examples-Tutorials-Recipes/Snapshots.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+the original `Statuses` cube in order to reuse the dimension definitions. We only need
+to add a new dimension that indicates the `date` of a snapshot. We're also referencing
+the definition of the `Statuses` cube with the
+[`sql()` function](https://cube.dev/docs/schema/reference/cube#parameters-sql).
```
Comment (hassankhan) [inline docs/content/Examples-Tutorials-Recipes/Snapshots.md:120]:
```suggestion
[`sql()` property](https://cube.dev/docs/schema/reference/cube#parameters-sql).
```

**Review by hassankhan — APPROVED** [review]
Minor nitpicks, great job @igorlukanin :raised_hands:

**Comment by keydunov** [comment]
It would be helpful to rebase it before review because it contains changes not related to snapshots.

File: examples/recipes/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+### Data sources
+
+- [Enable SSL Connections to the Database](https://cube.dev/docs/recipes/enable-ssl-connections-to-database)
+
+### Data schema
+
+- [Dynamically Union Tables](https://cube.dev/docs/recipes/dynamically-union-tables)
+- [Data Blending](https://cube.dev/docs/recipes/data-blending)
+- [Snapshots](/recipes/snapshots)
```
Comment (rchkv) [inline examples/recipes/README.md:24]: User will see a 404 page. The only way I found was to specify the full path

**Review by rchkv — APPROVED** [review]
Looks great! 👏

File: examples/recipes/README.md (hunk 4 above)
Comment (igorlukanin) [reply examples/recipes/README.md]: Good catch!

File: docs/content/Examples-Tutorials-Recipes/Snapshots.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+```javascript
+// Shipped as of May 1, 2019:
+[
+  {
+    "StatusSnapshots.count": 25
+  }
+]
+```
+
+## Source code
```
Comment (keydunov) [inline docs/content/Examples-Tutorials-Recipes/Snapshots.md:169]: I'll check the examples and will update you.

**Review by keydunov — CHANGES_REQUESTED** [review]
Looks good overall, left few comments.

File: docs/content/Examples-Tutorials-Recipes/Snapshots.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+```javascript
+cube(`StatusSnapshots`, {
+  extends: Statuses,
+
+  sql: `
+    -- Create a range from the earlist date to the latest date
+    WITH range AS (
+      SELECT date
+      FROM GENERATE_SERIES(
```
Comment (keydunov) [inline docs/content/Examples-Tutorials-Recipes/Snapshots.md:78]: this sql is unique to bigquery (and maybe supported by other databases), we need to mention it.

File: docs/content/Examples-Tutorials-Recipes/Snapshots.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+<!-- prettier-ignore-start -->
+[[info | ]]
+| We can consider the status property to be a
+| [slowly changing dimension](https://en.wikipedia.org/wiki/Slowly_changing_dimension)
+| (SCD) of type 2. Modeling data schemas with slowly changing dimensions is an
+| essential part of the data engineering skillset.
+<!-- prettier-ignore-end -->
+
+## Data schema
```
[the remark this answers is not in the document]
Comment (igorlukanin) [reply docs/content/Examples-Tutorials-Recipes/Snapshots.md:24]: Surely, a table would be better. However, I would rather not code tables in Markdown manually (too much labor). We're going to update the recipes with MDX components shortly. I think we will make the one that displays the results to support the table view.

File: docs/content/Examples-Tutorials-Recipes/Snapshots.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+We can see that statuses change occasionally. How do we count orders that remained
+in the `shipped` status at a particular date?
+
+First, we need to generate a range with all dates of interest, from the earliest to
+the latest. Second, we need to join the dates with the statuses and leave only the
+most recent statuses to date. 
+
+```javascript
+cube(`StatusSnapshots`, {
```
[the remark this answers is not in the document]
Comment (igorlukanin) [reply docs/content/Examples-Tutorials-Recipes/Snapshots.md:71]: I thought it's enough to describe them in text )above)...

> In the recipe below we'll learn, for a cube with `Product Id`, `Status`, and
> `Changed At` dimensions, how to calculate the snapshots of statuses at any given date.

...as well as to show them in the result set.

File: docs/content/Examples-Tutorials-Recipes/Snapshots.md (hunk 6 above)
Comment (igorlukanin) [reply docs/content/Examples-Tutorials-Recipes/Snapshots.md:78]: It's Postgres-specific, but I get your point.

**Comment by igorlukanin** [comment]
Closing in favor of #3295.

**Pull request closed without merging**